# -*- coding: utf-8 -*-
"""
Created on Fri May  4 12:15:56 2018

@author: morte
"""

configfile: "/home/masih/BacterialSimulator/Pipeline/ConfigFile.yaml"

rule vcfCaller:
    input:
        FastaSource = expand("{outputDIR}/{SequenceDIR}/{fasta_output}",outputDIR=config["outputDIR"],SequenceDIR=config["SequenceDIR"],fasta_output=str(config["fasta_file"]).split('/')[-1]),
        BamIndex= expand("{outputDIR}/{simulations}/{{simulation_index}}/Assembly/{{name}}.bam.mapq20.bai",outputDIR=config["outputDIR"],simulations=config["simulations"]),
        bamMapQFilter=expand("{outputDIR}/{simulations}/{{simulation_index}}/Assembly/{{name}}.bam.mapq20",outputDIR=config["outputDIR"],simulations=config["simulations"]),
    output:
        vcf=expand("{outputDIR}/{simulations}/{{simulation_index}}/VarCall/{{name}}.vcf",outputDIR=config["outputDIR"],simulations=config["simulations"]),
        Screentemps=temp(expand("{outputDIR}/{simulations}/{{simulation_index}}/VarCall/{{name}}.vcfScreentemps",outputDIR=config["outputDIR"],simulations=config["simulations"])),
    params:
        gatk=config['gatk'],
        logNAME="VCF caller." + strftime("%Y-%m-%d.%H-%M-%S", localtime()),
    run:
        from subprocess import call
        callString="%s --java-options \"-Xmx500M\" HaplotypeCaller -ploidy 1  -R %s -I %s -O %s &> %s" %(params.gatk,str(input.FastaSource),str(input.bamMapQFilter),str(output.vcf),str(output.Screentemps))
        call('echo "' + str(params.logNAME) + ':\n ' + callString + '\n" >> ' + config["ShellCallFile"], shell=True)
        call(callString, shell=True)