#writing the snakemake file to simulate the genome with simbac
# -*- coding: utf-8 -*-
"""
Created on Fri june 11

@author: Masih
?? ?? ?????? ????? ?????, ???? ??? ?? ????? ???

Simulate phylogentic tree

"""

# Used for system calls.
from subprocess import call

# Used for timestamping the log files.
from time import localtime, strftime
import os
#-----------------------------------------------------------------------------------
# SNAKEMAKE RULE #
   
rule simbac:
    input: 
    output: 
        phylogeny=expand("{outputDIR}/simulations/simbac_output.nwk",outputDIR=config["outputDIR"]),
        fasta=expand("{outputDIR}/simulations/simbac_output.fasta",outputDIR=config["outputDIR"]),
    params:
        num_species= config["num_species"]+1,
        genome_length= config["genome_length"],
        mutation_rate= config["mutation_rate"],
        r_i= config["r_i"],
        r_e=config["r_e"],
        random_seed= config['random_seed'],
        simbac_path=config['simbac_path'],
        outDir=config['outputDIR'],
        logNAME="Simulating genome and phylogeny." + strftime("%Y-%m-%d.%H-%M-%S", localtime()),
        simbac=os.path.join('modules','simbac','simbac.py'),
        shellCallFile=os.path.join(config["outputDIR"],'BacGWASim.log'),
    run:  
        #os.mkdir(os.path.join(params.outDir,'simulations'))
        callString='python3 %s --num_species %s --genome_length %s --output %s \
        --mutation_rate %s --r_i %s --r_e  %s --random_seed %s --simbac_path %s' % (params.simbac,params.num_species,params.genome_length,os.path.join(params.outDir,'simulations','simbac_output'),
                                                              params.mutation_rate,params.r_i,params.r_e,params.random_seed,params.simbac_path)
        call('echo "' + str(params.logNAME) + ':\n ' + callString + '\n" >> ' + params.shellCallFile, shell=True)
        call(callString, shell=True)
